internal/schema/to_from_nested_object_test.go

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package schema
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-plugin-codegen-spec/code"
+	"github.com/hashicorp/terraform-plugin-codegen-spec/schema"
+)
+
+func TestToFromNestedObject_renderFrom(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		name          string
+		assocExtType  *AssocExtType
+		fromFuncs     map[string]ToFromConversion
+		expected      []byte
+		expectedError error
+	}{
+		"default": {
+			name: "Example",
+			assocExtType: &AssocExtType{
+				AssociatedExternalType: &schema.AssociatedExternalType{
+					Type: "*apisdk.Type",
+				},
+			},
+			fromFuncs: map[string]ToFromConversion{
+				"bool_attribute": {
+					Default: "BoolPointerValue",
+				},
+			},
+			expected: []byte(`
+func (v ExampleValue) FromApisdkType(ctx context.Context, apiObject *apisdk.Type) (ExampleValue, diag.Diagnostics) {
+var diags diag.Diagnostics
+
+if apiObject == nil {
+return NewExampleValueNull(), diags
+}
+
+return ExampleValue{
+BoolAttribute: types.BoolPointerValue(apiObject.BoolAttribute),
+state: attr.ValueStateKnown,
+}, diags
+}
+`),
+		},
+		"nested-assoc-ext-type": {
+			name: "Example",
+			assocExtType: &AssocExtType{
+				AssociatedExternalType: &schema.AssociatedExternalType{
+					Type: "*apisdk.Type",
+				},
+			},
+			fromFuncs: map[string]ToFromConversion{
+				"bool_attribute": {
+					AssocExtType: &AssocExtType{
+						AssociatedExternalType: &schema.AssociatedExternalType{
+							Type: "*api.BoolAttribute",
+						},
+					},
+				},
+			},
+			expected: []byte(`
+func (v ExampleValue) FromApisdkType(ctx context.Context, apiObject *apisdk.Type) (ExampleValue, diag.Diagnostics) {
+var diags diag.Diagnostics
+
+if apiObject == nil {
+return NewExampleValueNull(), diags
+}
+
+boolAttributeVal, d := BoolAttributeValue{}.FromApiBoolAttribute(ctx, apiObject.BoolAttribute)
+
+diags.Append(d...)
+
+if diags.HasError() {
+return NewExampleValueNull(), diags
```

Comment:
For the `From*` functions, if there is an error, do we want to default to returning Unknown? Just to be sure there is an error if the provider dev doesn't check the diags 😆

Reply:
Good shout. I've updated List, Map, Object, and Set. I also updated for nested object (i.e., list, map, set, single nested attribute, and list, set, single nested block).